flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/HeapInternalTimerService.java

```diff
@@ -0,0 +1,325 @@
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.streaming.runtime.operators.Triggerable;
+import org.apache.flink.streaming.runtime.tasks.ProcessingTimeService;
+import org.apache.flink.util.InstantiationUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.HashSet;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link InternalTimerService} that stores timers on the Java heap.
+ */
+public class HeapInternalTimerService<K, N> implements InternalTimerService<N>, Triggerable {
+
+	private final TypeSerializer<K> keySerializer;
+
+	private final TypeSerializer<N> namespaceSerializer;
+
+	private final ProcessingTimeService processingTimeService;
+
+	private long currentWatermark = Long.MIN_VALUE;
+
+	private final org.apache.flink.streaming.api.operators.Triggerable<K, N> triggerTarget;
+
+	private final KeyContext keyContext;
+
+	/**
+	 * Processing time timers that are currently in-flight.
+	 */
+	private final PriorityQueue<InternalTimer<K, N>> processingTimeTimersQueue;
+	private final Set<InternalTimer<K, N>> processingTimeTimers;
+
+	protected ScheduledFuture<?> nextTimer = null;
+
+	/**
+	 * Currently waiting watermark callbacks.
+	 */
+	private final Set<InternalTimer<K, N>> watermarkTimers;
+	private final PriorityQueue<InternalTimer<K, N>> watermarkTimersQueue;
+
+	public HeapInternalTimerService(
+			TypeSerializer<K> keySerializer,
+			TypeSerializer<N> namespaceSerializer,
+			org.apache.flink.streaming.api.operators.Triggerable<K, N> triggerTarget,
+			KeyContext keyContext,
+			ProcessingTimeService processingTimeService) {
+		this.keySerializer = checkNotNull(keySerializer);
+		this.namespaceSerializer = checkNotNull(namespaceSerializer);
+		this.triggerTarget = checkNotNull(triggerTarget);
+		this.keyContext = keyContext;
+		this.processingTimeService = checkNotNull(processingTimeService);
+
+		watermarkTimers = new HashSet<>();
+		watermarkTimersQueue = new PriorityQueue<>(100);
+
+		processingTimeTimers = new HashSet<>();
+		processingTimeTimersQueue = new PriorityQueue<>(100);
+	}
+
+	public HeapInternalTimerService(
+			TypeSerializer<K> keySerializer,
+			TypeSerializer<N> namespaceSerializer,
+			org.apache.flink.streaming.api.operators.Triggerable<K, N> triggerTarget,
+			KeyContext keyContext,
+			ProcessingTimeService processingTimeService,
+			RestoredTimers<K, N> restoredTimers) {
+
+		this.keySerializer = checkNotNull(keySerializer);
+		this.namespaceSerializer = checkNotNull(namespaceSerializer);
+		this.triggerTarget = checkNotNull(triggerTarget);
+		this.keyContext = keyContext;
+		this.processingTimeService = checkNotNull(processingTimeService);
+
```

Comment:
That one's a bit tricky since I only get the `TypeSerializers` once the user requests a `TimerService`. Right now restoring is a two-step process, first the timers are read, and then when a user request the timer service are they actually put into a TimerService and we re-register the physical processing-time timer.